skorch/net.py

```diff
@@ -492,57 +502,52 @@ class Dummy:
             cb.initialize()
             callbacks_.append((name, cb))
 
+        # pylint: disable=attribute-defined-outside-init
         self.callbacks_ = callbacks_
-
         return self
 
     def initialize_criterion(self):
-        """Initializes the criterion."""
-        criterion_params = self.get_params_for('criterion')
-        self.criterion_ = self.criterion(**criterion_params)
-        if isinstance(self.criterion_, torch.nn.Module):
-            self.criterion_ = to_device(self.criterion_, self.device)
-        return self
+        """Initializes the criterion.
 
-    def _format_reinit_msg(self, name, kwargs=None, triggered_directly=True):
-        """Returns a message that informs about re-initializing a compoment.
-
-        Sometimes, the module or optimizer need to be
-        re-initialized. Not only should the user receive a message
-        about this but also should they be informed about what
-        parameters, if any, caused it.
+        If the criterion is already initialized and no parameter was changed, it
+        will be left as is.
 
         """
-        msg = "Re-initializing {}".format(name)
-        if triggered_directly and kwargs:
-            msg += (" because the following parameters were re-set: {}."
-                    .format(', '.join(sorted(kwargs))))
-        else:
-            msg += "."
-        return msg
+        kwargs = self.get_params_for('criterion')
+        criterion = self.criterion
+        is_init = isinstance(criterion, torch.nn.Module)
+        if is_init and not kwargs:
+            # criterion already initialized and no params changed:
+            self.criterion_ = self.criterion
+            return
+
+        if is_init:
+            criterion = type(criterion)
+
+        # pylint: disable=attribute-defined-outside-init
+        self.criterion_ = criterion(**kwargs)
+        return self
```

Comment:
Can't we refactor the initialization check into its own function?

```python
def initialized_instance(instance_or_cls, kwargs):
    is_init = isinstance(instance_or_cls, torch.nn.Module):
    if is_init and not kwargs:
        return instance_or_cls
    if is_init:
        return type(instance_or_cls)
    return instance_or_cls(**kwargs)
```

`initialize_criterion` might look like this then:
```
kwargs = self.get_params_for('criterion')
self.criterion_ = initialized_instance(self.criterion, kwargs)
```

A user who wants full control may opt to not use it but it would make the code a lot more compact at the cost of removing transparency.

Reply:
I agree, this is a worthwhile tradeoff.